### PR TITLE
Capture non tensor arguments in record_function

### PR DESCRIPTION
**Pull request by shengfukevin**
Summary: RECORD_FUNCTION only capture the argument when it is a Tensor. However, it is very common for user to use the argument with primitive data type (int, float, index, bool). This DIFF is to support non tensor arguments in RECORD_FUNCTION.

Test Plan:
unit test
    buck test  mode/dev-nosan caffe2/test:profiler -- test_execution_trace_with_pt2 test_execution_trace_alone test_execution_trace_with_kineto test_execution_trace_start_stop test_execution_trace_repeat_in_loop test_execution_trace_no_capture

Differential Revision: D53674768

### CODE REVIEW

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -850,6 +853,9 @@ std::pair<UnpackedInput, InputFlags> unpack_input(PyObject* args) {
   auto num_args = PyTuple_GET_SIZE(args);
   unpacked.input_tuple = PyTuple_New(num_args);
   flags.needs_input_grad = PyTuple_New(num_args);
+  bool need_input = torch::autograd::profiler::profilerEnabled() &&
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp:856]:
```suggestion
  bool profiler_need_input = torch::autograd::profiler::profilerEnabled() &&
```

File: torch/csrc/autograd/python_function.cpp [hunk 2]
```diff
@@ -865,12 +871,23 @@ std::pair<UnpackedInput, InputFlags> unpack_input(PyObject* args) {
       }
       Py_INCREF(Py_False);
       PyTuple_SET_ITEM(flags.needs_input_grad.get(), i, Py_False);
+
+      if (need_input) {
+        // The following conversion from PyObject to IValue is expensive
+        // Only do it if profiler is enabled and needs input shapes
+        auto match = torch::jit::tryToInferType(arg);
+        if (match.success()) {
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp:879]: If the idea here to want to record ALL arguments so you can replay it later, what happens later when there's a type you cannot infer?

File: torch/csrc/autograd/python_function.cpp [hunk 3]
```diff
@@ -865,12 +871,23 @@ std::pair<UnpackedInput, InputFlags> unpack_input(PyObject* args) {
...
+          unpacked.input_arguments.push_back(
+              torch::jit::toIValue(arg, match.type()));
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp:881]: std::move?

File: torch/csrc/autograd/python_function.cpp [hunk 4]
```diff
@@ -833,6 +834,8 @@ static std::unordered_set<at::TensorImpl*> _parse_non_differentiable(
 struct UnpackedInput {
   THPObjectPtr input_tuple;
   variable_list input_vars;
+  // input_arguments is for RECORD_FUNCTION only
+  std::vector<c10::IValue> input_arguments;
```
Comment (soulitzer) [inline torch/csrc/autograd/python_function.cpp:838]:
```suggestion
  std::vector<c10::IValue> record_function_inputs;
```

File: torch/csrc/autograd/python_function.cpp (hunk 2 above)
Comment (shengfukevin) [reply torch/csrc/autograd/python_function.cpp:879]: The type that can not be inferred are nn.Module and ScriptModule:
1. https://www.internalfb.com/code/fbsource/[ef3e69d17cea50617109f1c9b47d29db0891b443]/fbcode/caffe2/torch/csrc/jit/python/pybind_utils.h?lines=489
2. https://www.internalfb.com/code/fbsource/[ef3e69d17cea50617109f1c9b47d29db0891b443]/fbcode/caffe2/torch/csrc/jit/python/pybind_utils.h?lines=495

I do not see a use case that an operator need these types to run. The use case I want to support is having some primitive data types or their container to define an op. For example, padding and strides in a convolution op.

File: torch/csrc/autograd/python_function.cpp (hunk 3 above)
Comment (shengfukevin) [reply torch/csrc/autograd/python_function.cpp:881]: Since IValue here is a temporary variable, I do not think it needs std::move. I got a compile error if I add that:
moving a temporary object prevents copy elision

File: torch/csrc/autograd/python_function.cpp (hunk 4 above)
Comment (shengfukevin) [reply torch/csrc/autograd/python_function.cpp:838]: Fixed

File: torch/csrc/autograd/python_function.cpp (hunk 1 above)
Comment (shengfukevin) [reply torch/csrc/autograd/python_function.cpp:856]: Fixed.

File: torch/csrc/autograd/python_function.cpp (hunk 2 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_function.cpp:879]: Do you have a code pointer to how this is normally done? I think we want to make sure they do the same thing.

**Review by soulitzer — APPROVED** [review]

**Comment by awgu** [comment]
Hi @shengfukevin!

With your PR, I am seeing behavior like this when running our new FSDP implementation that uses a custom autograd function with non-tensor input:
Screenshot 2024-02-28 at 2 04 41 PM
in my custom autograd function, and it is causing significant CPU overhead.

I checked that by reverting this PR locally, the issue goes away. I wanted to get your thoughts here. It seems like profiling is now causing some jit script to run.

**Comment by awgu** [comment]
I further verified that this expensive CPU overhead happens on every profiled iteration, not just the 1st one. In other words, it is not a 1-time cost.

Screenshot 2024-02-28 at 3 12 26 PM

**Comment by soulitzer** [comment]
Thanks for the report @awgu, this overhead is because the type of the IValue cannot be inferred in the custom Function case and so there needs to be additional overhead when profiling. 
@shengfukevin I will revert this because we don't want to regress perf for existing profiling settings.

This seems hard to avoid because as mentioned above, we probably need to do some extra overhead in the custom Function  to do the inference, but in the reland, we should at least hide this new feature behind a flag (unless we can do something smarter about reducing/amortizing away the overhead here).

**Comment by awgu** [comment]
One thing to investigate is why in the FSDP repro, the expensive JIT script logic runs on multiple profiled iterations, not just the first. In a simple local repro, I do not see this behavior.

<details>

```
from typing import Callable

import torch


def benchmark_with_profiler(
    benchmark_fn: Callable,
    *benchmark_fn_args,
    **benchmark_fn_kwargs,
) -> None:
    wait, warmup, active = 1, 1, 5
    num_steps = wait + warmup + active
    rank = 0
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        schedule=torch.profiler.schedule(
            wait=wait, warmup=warmup, active=active, repeat=1, skip_first=1
        ),
        on_trace_ready=torch.profiler.tensorboard_trace_handler("./"),
        record_shapes=True,
        profile_memory=True,
        with_stack=True,
        with_flops=True,
        with_modules=False,
    ) as prof:
        for step_idx in range(1, num_steps + 1):
            benchmark_fn(*benchmark_fn_args, **benchmark_fn_kwargs)
            if rank is None or rank == 0:
                prof.step()


class MyFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x: torch.Tensor, w: torch.Tensor, obj) -> torch.Tensor:
        return x @ w


x = torch.randn((1024, 1024), device="cuda")
w = torch.randn((1024, 1024), device="cuda")

class A:
    ...

obj = A()

def inner():
    with torch.profiler.record_function("x @ w"):
        y = MyFunction.apply(x, w, obj)

benchmark_with_profiler(inner)
```

</details>

Screenshot 2024-02-28 at 3 49 17 PM

**Comment by awgu** [comment]
I shared a repro with FSDP with @shengfukevin on chat but adding here for completeness:

<details>

```
import os

import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed._composable.fsdp import fully_shard


def benchmark_with_profiler(
    benchmark_fn,
    *benchmark_fn_args,
    **benchmark_fn_kwargs,
) -> None:
    wait, warmup, active = 1, 1, 3
    num_steps = wait + warmup + active
    rank = dist.get_rank()
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
        schedule=torch.profiler.schedule(
            wait=wait, warmup=warmup, active=active, repeat=1, skip_first=1
        ),
        on_trace_ready=torch.profiler.tensorboard_trace_handler("./")
        if rank == 0
        else None,
        record_shapes=True,
        profile_memory=True,
        with_stack=True,
        with_flops=True,
        with_modules=False,
    ) as prof:
        for step_idx in range(1, num_steps + 1):
            benchmark_fn(*benchmark_fn_args, **benchmark_fn_kwargs)
            if rank is None or rank == 0:
                prof.step()


dist.init_process_group(backend="nccl")
gpu_id = int(os.environ["LOCAL_RANK"])
device = f"cuda:{gpu_id}"
torch.cuda.set_device(device)

model = nn.Sequential(nn.Linear(3, 3), nn.Linear(3, 3), nn.Linear(3, 3))
for lin in model:
    fully_shard(lin)
fully_shard(model)

inp = torch.randn((1, 3), device="cuda")


def inner():
    model(inp).sum().backward()


benchmark_with_profiler(inner)
```

```
torchrun --standalone --nproc_per_node=2 repro_jit_script.py 
```

</details>
